lib/common/db-renderable-content.js

```diff
@@ -70,26 +82,84 @@ function dbRenderableContentServiceFactory(
         });
     };
 
-    DbRenderableContentService.prototype.put = function (file, contents, scope) {
+    DbRenderableContentService.prototype.loadFile = function(filename, filepath, content, scope) {
         var self = this;
         scope = scope || Constants.Scope.Global;
-        return waterline[self.collectionName].findOne({ name: file, scope: scope })
-            .then(function(doc) {
-                if (!_.isEmpty(doc)) {
-                    return waterline[self.collectionName].update(
-                        { name: file, scope: scope },
-                        { contents :contents }
-                    );
+
+        return Promise.try(function() {
+            return crypto.createHash('md5').update(content).digest('base64');
+        })
+        .then(function(hash) {
+            return waterline[self.collectionName].findOne({name: filename, scope: scope})
+            .then(function(fileRecord) {
+                if (_.isEmpty(fileRecord)) {
+                    return waterline[self.collectionName].create({
+                        name: filename,
+                        hash: hash,
+                        path: filepath,
+                        scope: scope
+                    });
                 } else {
-                    return waterline[self.collectionName].create(
-                        {
-                            name: file,
-                            contents: contents,
-                            scope: scope
-                        }
-                    );
+                    if (fileRecord.hash !== hash) {
+                        fileRecord.hash = hash;
+                        fileRecord.path = filepath;
+                        return fileRecord.save();
+                    }
+                    return fileRecord;
                 }
             });
+        });
+    };
+
+    DbRenderableContentService.prototype.put = function (filename, stream, scope) {
+        var self = this;
+        var contents = '';
+        var resolve;
+        var reject;
+        var filepath = path.join(self.directory, filename);
+        scope = scope || Constants.Scope.Global;
+
+        var promise = new Promise(function(_resolve, _reject) {
+            resolve = _resolve;
+            reject = _reject;
+        });
+
+        stream.on('data', function(chunk) {
+            contents += chunk.toString('utf-8');
```

Comment:
Could the hash be computed here and the hash passed into loadFile() to avoid reading the file contents twice?